collect/cache/cuckoo.go

```diff
@@ -59,26 +62,46 @@ func NewCuckooTraceChecker(capacity uint, m metrics.Metrics) *CuckooTraceChecker
 		future:   nil,
 		met:      m,
 		addch:    make(chan string, AddQueueDepth),
+		done:     make(chan struct{}),
 	}
 	for _, metric := range cuckooTraceCheckerMetrics {
 		m.Register(metric)
 	}
 
 	// To try to avoid blocking on Add, we have a goroutine that pulls from a
 	// channel and adds to the filter.
+	c.shutdownWG.Add(1)
 	go func() {
+		defer c.shutdownWG.Done()
+
 		ticker := time.NewTicker(AddQueueSleepTime)
-		for range ticker.C {
-			// as long as there's anything still in the channel, keep trying to drain it
-			for len(c.addch) > 0 {
-				c.drain()
+		for {
+			select {
+			case <-ticker.C:
+				for len(c.addch) > 0 {
+					c.drain()
+				}
+			case <-c.done:
+				return
+
 			}
 		}
 	}()
 
 	return c
 }
 
+func (c *CuckooTraceChecker) Stop() {
+	// stop the goroutine that drains the add channel
+	close(c.done)
```

Comment:
You should close c.addch here also. Assuming everything is shutting down in the correct order, you shouldn't get any more calls to Add(). If things are not shutting down, you'll find out because it will panic!

Note doing so means you'll also need to use the two-value channel read in drain to detect if it's closed.